blog: DIY: Create Your Own Cloud with Kubernetes (Part 2) blog post (#45432)
kubernetes/website: content/en/blog/_posts/2024-04-05-diy-create-your-own-cloud-with-kubernetes-part-2/index.md

---
layout: blog
title: "DIY: Create Your Own Cloud with Kubernetes (Part 2)"
slug: diy-create-your-own-cloud-with-kubernetes-part-2

> date: 2024-04-05

Reviewer: Let's pick a particular time during this day for the article to publish. They ought to be at least 5 minutes apart, and in order.

At Ænix, we wanted to provide a ready-to-use solution that could be used without the need to purchase and setup an additional external storage, and that was optimal in terms of speed and resource utilization. LINSTOR became that solution.
The time-tested and industry-popular technologies such as LVM and ZFS as backend gives confidence that data is securely stored. DRBD-based replication is incredible fast and consumes a small amount of computing resources.

For installing LINSTOR in Kubernetes, there is the Piraeus project, which already provides a ready-made block storage to use with KubeVirt.

{{< note >}}
In case you are using Talos Linux, as we described in the [previous article](/blog/2024/04/05/diy-create-your-own-cloud-with-kubernetes-part-1/), you will need to enable the necessary kernel modules in advance, and configure piraeus as described in the [instruction](https://github.com/piraeusdatastore/piraeus-operator/blob/v2/docs/how-to/talos.md).
{{< /note >}}

## Networking for Kubernetes VMs

> Despite having the similar interface - CNI, The network architecture in Kubernetes is actually more complex and typically consists of many independent components that are not directly connected to each other. In fact you can split Kubernetes networking into four layers, which are described below.

Reviewer 2: ```suggestion
Despite having the similar interface - CNI, The network architecture in Kubernetes is actually more complex and typically consists of many independent components that are not directly connected to each other. In fact, you can split Kubernetes networking into four layers, which are described below.
```

One additional aspect to keep in mind when running virtual machines in Kubernetes is the need for IPAM (IP Address Management), especially for secondary interfaces provided by Multus. This is commonly managed by a DHCP server operating within your infrastructure. Additionally, the allocation of MAC addresses for virtual machines can be managed by [Kubemacpool](https://github.com/k8snetworkplumbingwg/kubemacpool).

Although in our platform, we decided to go another way and fully rely on [Kube-OVN](https://www.kube-ovn.io/). This CNI plugin is based on OVN (Open Virtual Network) which was originally developed for OpenStack and it provides a complete network solution for virtual machines in Kubernetes, features Custom Resources for managing IPs and MAC addresses, supports live migration with preserving IP addresses between the nodes, and enables the creation of VPCs for physical network separation between tenants.

In Kube-OVN you can assign separate subnets to an entire namespace or connect them as additional network interfaces using Multus.

### Services Network

In addition to the CNI plugin, Kubernetes also has a services network, which is primarily needed for service discovery.

> Contrary to traditional virtual machines, Kubernetes originally designed to run pods with a random address.

Reviewer 2: ```suggestion
Contrary to traditional virtual machines, Kubernetes is originally designed to run pods with a random address.
```

And the services network provides a convenient abstraction (stable IP addresses and DNS names) that will always direct traffic to the correct pod.
The same approach is also commonly used with virtual machines in clouds despite the fact that their IPs are usually static.

{{< figure src="net-services.svg" caption="A diagram showing the role of the services network (services network plugin) on the Kubernetes network scheme" alt="A diagram showing the role of the services network (services network plugin) on the Kubernetes network scheme" >}}

The implementation of the services network in Kubernetes is handled by the services network plugin, The standard implementation is called **kube-proxy** and is used in most clusters.
But nowadays, this functionality might be provided as part of the CNI plugin. The most advanced implementation is offered by the [Cilium](https://cilium.io/) project, which can be run in kube-proxy replacement mode.

Cilium is based on the eBPF technology, which allows for efficient offloading of the Linux networking stack, thereby improving performance and security compared to traditional methods based on iptables.

> In practice, Cilium and Kube-OVN can be easily [integrated](https://kube-ovn.readthedocs.io/zh-cn/stable/en/advance/with-cilium/) to provide a unified solution that offers seamless multi-tenant networking for virtual machines, as well as advanced network policies and combined services network functionality.

Reviewer 2: ```suggestion
In practice, Cilium and Kube-OVN can be easily [integrated](https://kube-ovn.readthedocs.io/zh-cn/stable/en/advance/with-cilium/) to provide a unified solution that offers seamless, multi-tenant networking for virtual machines, as well as advanced network policies and combined services network functionality.
```

For bare metal Kubernetes clusters, there are several load balancers available: [MetalLB](https://metallb.universe.tf/), [kube-vip](https://kube-vip.io/), [LoxiLB](https://www.loxilb.io/), also [Cilium](https://docs.cilium.io/en/latest/network/lb-ipam/) and [Kube-OVN](https://kube-ovn.readthedocs.io/zh-cn/latest/en/guide/loadbalancer-service/) provides built-in implementation.

The role of a external load balancer is to provide a stable address available externally and direct external traffic to the services network.
The services network plugin will direct it to your pods and virtual machines as usual.

{{< figure src="net-services.svg" caption="A diagram showing the role of the external load balancer on the Kubernetes network scheme" alt="The role of the external load balancer on the Kubernetes network scheme" >}}

In most cases, setting up a load balancer on bare metal is achieved by creating floating IP address on the nodes within the cluster, and announce it externally using ARP/NDP or BGP protocols.

After exploring various options, we decided that MetalLB is the simplest and most reliable solution, although we do not strictly enforce the use of only it.

> Another benefit is that in L2 mode, MetalLB speakers continuously checking thir neighbors state by sending preforming liveness checks using a memberlist protocol.

Reviewer 2: ```suggestion
Another benefit is that in L2 mode, MetalLB speakers continuously check their neighbour's state by sending preforming liveness checks using a memberlist protocol.
```

This enables failover that works independently of Kubernetes control-plane.

## Conclusion

> This concludes our overview of virtualization, storage, and networking in Kubernetes. The technologies mentioned here are available and already pre-configured on the [Cozystack](https://github.com/aenix-io/cozystack) platform, where you can try them with no limitations.

Reviewer 2: Does 'our' in this context refer to a group of contributors?

Author: Yes. Group of contributors or authors to the article. Especially me :)

> In the [next article](/blog/2024/04/05/diy-create-your-own-cloud-with-kubernetes-part-3/), we will detail how, on top of this, we implement the provisioning of fully functional Kubernetes clusters with just the push of a button.

Reviewer 2: Does 'we' in this context, refer to a group of individuals?

We avoid using 'we' for readers. Rather, you should use 'you' for them.

Reviewer: Try:
```diff
-In the [next article](/blog/2024/04/05/diy-create-your-own-cloud-with-kubernetes-part-3/), we will detail how, on top of this, we implement the provisioning of fully functional Kubernetes clusters with just the push of a button.
+In the [next article](/blog/2024/04/05/diy-create-your-own-cloud-with-kubernetes-part-3/), I'll
+detail how, on top of this, you can implement the provisioning of fully functional Kubernetes clusters
+with just the push of a button.
```

(the “we” becomes an “I” and a ”you”, which helps readers especially if English is not their first language)

Author: Thanks, accepted
